### ISSUE TITLE
regression: storaged crash loop: some element of the Sink as_of frontier is too far advanced for our output-gating timestamp"

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#14754]
### What version of Materialize are you using?

78a4e9e95d319e224af4b2df20c19591de054d4a

### How did you install Materialize?

Docker image

### What is the issue?

The Zippy kafka sources test has started failing . The test creates a sink based on a materialized view based on one or more Kafka sources. The test also kills and restarts storageds

https://buildkite.com/materialize/nightlies/builds/1147#01832df6-7a81-4cce-8c7a-6e11dbbef095

The logs indicate that the panic below has happened 64 times, so it is a crash loop type of situation.

### Relevant log output

```source
zippy-materialized-1     | thread 'timely:work-0' panicked at 'some element of the Sink as_of frontier is too far advanced for our output-gating timestamp: as_of Antichain { elements: [1662896751762] }, gate_ts: 1662896751761', src/storage/src/sink/kafka.rs:1157:25
zippy-materialized-1     | stack backtrace:
zippy-materialized-1     |    0: rust_begin_unwind
zippy-materialized-1     |              at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
zippy-materialized-1     |    1: core::panicking::panic_fmt
zippy-materialized-1     |              at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
zippy-materialized-1     |    2: mz_storage::sink::kafka::produce_to_kafka::{{closure}}::{{closure}}::{{closure}}
zippy-materialized-1     |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-02dcec3a2b709fcad-1/materialize/tests/src/storage/src/sink/kafka.rs:1157:25
zippy-materialized-1     |    3: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
zippy-materialized-1     |              at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/future/mod.rs:91:19
zippy-materialized-1     |    4: mz_storage::sink::kafka::produce_to_kafka::{{closure}}::{{closure}}
zippy-materialized-1     |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-02dcec3a2b709fcad-1/materialize/tests/src/timely-util/src/operators_async_ext.rs:201:36
zippy-materialized-1     |    5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
zippy-materialized-1     |              at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/future/mod.rs:91:19
zippy-materialized-1     |    6: <futures_util::future::future::fuse::Fuse<Fut> as core::future::future::Future>::poll
zippy-materialized-1     |              at ./cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.24/src/future/future/fuse.rs:86:37
zippy-materialized-1     |    7: <core::pin::Pin<P> as core::future::future::Future>::poll
zippy-materialized-1     |              at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/future/future.rs:124:9
zippy-materialized-1     |    8: <timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G> as mz_timely_util::operators_async_ext::OperatorBuilderExt<G>>::build_async::{{closure}}::{{closure}}
zippy-materialized-1     |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-02dcec3a2b709fcad-1/materialize/tests/src/timely-util/src/operators_async_ext.rs:165:43
zippy-materialized-1     |    9: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/dataflow/operators/generic/builder_rc.rs:162:26
zippy-materialized-1     |   10: <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/dataflow/operators/generic/builder_raw.rs:204:9
zippy-materialized-1     |   11: timely::progress::subgraph::PerOperatorState<T>::schedule
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:684:30
zippy-materialized-1     |   12: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:344:26
zippy-materialized-1     |   13: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:310:17
zippy-materialized-1     |   14: timely::progress::subgraph::PerOperatorState<T>::schedule
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:684:30
zippy-materialized-1     |   15: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:344:26
zippy-materialized-1     |   16: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/progress/subgraph.rs:310:17
zippy-materialized-1     |   17: timely::worker::Wrapper::step::{{closure}}
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/worker.rs:754:57
zippy-materialized-1     |   18: core::option::Option<T>::map
zippy-materialized-1     |              at ./rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/option.rs:929:29
zippy-materialized-1     |   19: timely::worker::Wrapper::step
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/worker.rs:754:26
zippy-materialized-1     |   20: timely::worker::Worker<A>::step_or_park
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/worker.rs:396:38
zippy-materialized-1     |   21: mz_storage::storage_state::Worker<A>::run_client
zippy-materialized-1     |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-02dcec3a2b709fcad-1/materialize/tests/src/storage/src/storage_state.rs:134:17
zippy-materialized-1     |   22: mz_storage::storage_state::Worker<A>::run
zippy-materialized-1     |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-02dcec3a2b709fcad-1/materialize/tests/src/storage/src/storage_state.rs:110:33
zippy-materialized-1     |   23: mz_storage::protocol::server::serve::{{closure}}
zippy-materialized-1     |              at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-02dcec3a2b709fcad-1/materialize/tests/src/storage/src/protocol/server.rs:94:9
zippy-materialized-1     |   24: timely::execute::execute::{{closure}}
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/timely/src/execute.rs:287:22
zippy-materialized-1     |   25: timely_communication::initialize::initialize_from::{{closure}}
zippy-materialized-1     |              at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/1f3a315/communication/src/initialize.rs:316:33
zippy-materialized-1     | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
zippy-materialized-1     | environmentd: 2022-09-11T11:47:49.424093Z ERROR mz_orchestrator_process: storage-u50 exited: Ok(ExitStatus(unix_wait_status(6))); relaunching in 5s
```

**Comment by aljoscha** [comment]
Linking as a possibly related issue: https://github.com/MaterializeInc/materialize/issues/14728

**Comment by philip-stoev** [comment]
Happened again https://buildkite.com/materialize/nightlies/builds/1175#_

**Comment by aljoscha** [comment]
I believe I found the reason for this, fix incoming.

**Comment by aljoscha** [comment]
@bkirwi & @cjubb39 Turns out this is a somewhat subtle and tricky bug. I added more logging statements in https://github.com/MaterializeInc/materialize/pull/14902. You can then reproduce the bug with

```
$ bin/mzcompose --find zippy run default --scenario=KafkaSources --actions=200
```

Once it hangs, you can inspect the logs using `docker logs zippy-materialized-1` and will find sth like

```
storage-u8: 2022-09-20T23:06:40.658023Z  INFO mz_storage::sink::kafka: kafka-u8: initial as_of: Antichain { elements: [1663715199936] }
storage-u8: 2022-09-20T23:06:40.658044Z  INFO mz_storage::sink::kafka: Identified latest progress record: None
storage-u8: 2022-09-20T23:06:40.658059Z  INFO mz_storage::sink::kafka: kafka-u8: downgrading write frontier to: 0
storage-u8: 2022-09-20T23:06:40.659113Z  INFO mz_storage::sink::kafka: kafka-u8: sending progress for gate ts: 1663715199935
storage-u8: 2022-09-20T23:06:40.726746Z  INFO mz_persist_client::internal::machine: snapshot s45b57729-c5d1-4ccb-a697-5031dc928b4a as of Antichain { elements: [1663715199936] } not yet available for upper Antichain { elements: [1663715198001] } retrying in 512ms
storage-u8: 2022-09-20T23:06:41.150319Z  INFO mz_storage::sink::kafka: kafka-u8: downgrading write frontier to: 1663715199935
storage-u8: 2022-09-20T23:06:41.150600Z  INFO mz_storage::sink::kafka: kafka-u8: downgrading write frontier to: 1663715199935
storage-u8: 2022-09-20T23:06:41.150620Z  INFO mz_storage::sink::kafka: kafka-u8: downgrading write frontier to: 1663715199935
```

and later

```
storage-u8: 2022-09-20T23:06:53.660327Z  INFO mz_storage::sink::kafka: kafka-u8: initial as_of: Antichain { elements: [1663715199936] }
storage-u8: 2022-09-20T23:06:53.660354Z  INFO mz_storage::sink::kafka: Identified latest progress record: Some(1663715199935)
thread 'timely:work-0' panicked at 'kafka-u8: some element of the Sink as_of frontier is too far advanced for our output-gating timestamp: as_of Antichain { elements: [1663715199936] }, gate_ts: 1663715199935', src/storage/src/sink/kafka.rs:1163:25
```

Naively, I'd say we can only commit to Kafka once we're safely beyond the `as_of`, such that we can restart. Devil is in the details but I'm too tired now... 😅
